ir/visitor.cpp

```diff
@@ -19,54 +19,69 @@ limitations under the License.
-    // unfortunatelly maps use pairs all over the place, which is where they come from.
-    typedef unordered_map<const IR::Node *, std::pair<bool, const IR::Node *>>  visited_t;
+    struct visit_info_t {
+        bool visit_in_progress;
+        const IR::Node *result;
+    };
+    typedef unordered_map<const IR::Node *, visit_info_t>  visited_t;
     visited_t           visited;
 
  public:
-    struct change_t {
-        bool                                    valid;
-        std::pair<visited_t::iterator, bool>    state;  // result of visited.emplace
-        change_t() : valid(false) {}
-        change_t(visited_t *visited, const IR::Node *n) : valid(true),
-            state(visited->emplace(n, std::make_pair(false, n))) {
-                if (!state.second && !state.first->second.first)
-                    BUG("IR loop detected "); }
-        explicit operator bool() { return valid; }
-        bool done() { return !state.second; }
-        const IR::Node *orig() { return state.first->first; }
-        const IR::Node *result() { return state.first->second.second; }
-    };
-    bool done(const IR::Node *n) const { return visited.count(n) && visited.at(n).first; }
-    const IR::Node *result(IR::Node *n) const { return visited.at(n).second; }
-    change_t track(const IR::Node *n) { return change_t(&visited, n); }
-    void start(change_t &change) { change.state.first->second.first = false; }
-    bool finish(change_t &change, const IR::Node *orig, const IR::Node *final) {
-        if (!change.valid || (change.state.first = visited.find(orig)) == visited.end())
+
+    /* Begin tracking `n` during a visiting pass.  Use `finish` to mark `n` as
+     * visited once the pass completes.
+    */
+    void start(const IR::Node *n) {
+        // Initialization
+        visited_t::iterator visited_it;
+        bool inserted;
+        std::tie(visited_it, inserted) =
+            visited.emplace(n, visit_info_t({true,n}));
+
+        // Sanity check for IR loops
+        bool already_present = !inserted;
+        visit_info_t *visit_info = &(visited_it->second);
+        if (already_present && visit_info->visit_in_progress)
+            BUG("IR loop detected "); 
+    }
+
+    /* Mark `n` as finished.  `done(n)` will return true, and `result(n)` will
+     * return the resulting node, if any. */
+    bool finish(const IR::Node *orig, const IR::Node *final) {
+        auto it = visited.find(orig);
+        if (it == visited.end())
             BUG("visitor state tracker corrupted");
-        change.state.first->second.first = true;
+
+        visit_info_t *visit_info = &(it->second);
+        visit_info->visit_in_progress = false;
         if (!final || *final != *orig) {
-            change.state.first->second.second = final;
-            visited.emplace(final, std::make_pair(true, final));
+            visit_info->result = final;
+            visited.emplace(final, visit_info_t({false,final}));
```

Comment:
Good idea.  The constructor is called twice, once in `start` (the normal case) and once in `finish` (when the result is a brand new IR node).